trl/trainer/online_dpo_trainer.py

```diff
@@ -0,0 +1,619 @@
+        # pg_clipfrac_stats = torch.zeros(stats_shape, device=device)
+        # vf_loss_stats = torch.zeros(stats_shape, device=device)
+        # vf_clipfrac_stats = torch.zeros(stats_shape, device=device)
+        # entropy_stats = torch.zeros(stats_shape, device=device)
+        # ratio_stats = torch.zeros(stats_shape, device=device)
+        model.train()
+        self.state.max_steps = args.num_updates * args.num_mini_batches
+        self.state.num_train_epochs = args.total_episodes / self.train_dataset_len
+        self.state.is_local_process_zero = self.is_local_process_zero()
+        self.state.is_world_process_zero = self.is_world_process_zero()
+
+        # Compute absolute values for logging, eval, and save if given as ratio
+        if args.logging_steps is not None:
+            if args.logging_steps < 1:
+                self.state.logging_steps = math.ceil(self.state.max_steps * args.logging_steps)
+            else:
+                self.state.logging_steps = args.logging_steps
+        if args.eval_steps is not None:
+            if args.eval_steps < 1:
+                self.state.eval_steps = math.ceil(self.state.max_steps * args.eval_steps)
+            else:
+                self.state.eval_steps = args.eval_steps
+        if args.save_steps is not None:
+            if args.save_steps < 1:
+                self.state.save_steps = math.ceil(self.state.max_steps * args.save_steps)
+            else:
+                self.state.save_steps = args.save_steps
+
+        self.control = self.callback_handler.on_train_begin(args, self.state, self.control)
+        saved_data = {"prompt": [], "chosen": [], "rejected": [], "update": []}
+
+        for update in range(1, args.num_updates + 1):
+            self.state.episode += 1 * args.batch_size
+            self.lr_scheduler.step()
+            data = next(iter_dataloader)
+            with torch.no_grad():
+                queries = data["input_ids"].to(device)
+                context_length = queries.shape[1]
+                query_responses = []
+                responses = []
+                postprocessed_responses = []
+                logprobs = []
+                ref_logprobs = []
+                scores = []
+                sequence_lengths = []
+                with unwrap_model_for_generation(model, self.accelerator) as unwrapped_model:
+                    for i in range(0, queries.shape[0], args.local_rollout_forward_batch_size):
+                        query = queries[i : i + args.local_rollout_forward_batch_size]
+                        query_response, logits = generate(
+                            unwrapped_model,
+                            query,
+                            tokenizer.pad_token_id,
+                            generation_config,
+                        )
+                        response = query_response[:, context_length:]
+                        logits /= args.temperature + 1e-7
```

Comment:
You don't need to adjust the logits by temperature here because it's already adjusted in the `generate` function. 

```suggestion
```

Reply:
Does this mean I don't need to adjust it in other cases too? Or just here?